File: mfd/src/fitness/MFDFitnessFunction.scala
package fitness

import genotype.Genotype
import scala.io.Source
import collection.{SortedSet, mutable}


class MFDFitnessFunction(matrixFileName: String,
                         nSympt: Int, nDis: Int) extends FitnessFunction {

  var priorProb = Array.ofDim[Double](nDis)
  var tendencyMatrix = Array.ofDim[Double](nSympt, nDis)
  var mPlus = null.asInstanceOf[Set[Int]]
  var fitnessMap = new mutable.HashMap[Long, Double]()
  readTendencyMatrix()

  def calculateFitness(g: Genotype): Double = {
    var diagnosis = getSetIndices(g.individual, nDis)

    var l1 = L1(diagnosis)
    var l2 = L2(diagnosis)
    var l3 = L3(diagnosis)
    l1 * l2 * l3
    //    println("L1: " + l1 + "\tL2: " + l2 + "\tL3: " + l3)
  }

  def calculateFitness(diagInt: Long): Double = {
    var diagnosis = getSetIndices(diagInt, nDis)

    var l1 = L1(diagnosis)
    var l2 = L2(diagnosis)
    var l3 = L3(diagnosis)
    l1 * l2 * l3
    //    println("L1: " + l1 + "\tL2: " + l2 + "\tL3: " + l3)
  }

  def setMPlus(mPlusNum: Int) {
    mPlus = getSetIndices(mPlusNum, nSympt)
//    println(mPlusNum.toBinaryString + "\t" + mPlus)
  }

  def getSetIndices(num: Long, binStrLen: Int): Set[Int] = {
    var s = num.toBinaryString
    var res = Set[Int]()

    for (i <- 0 until s.length) {
      if (s(i) == '1') {
        res += (binStrLen - s.length + i)
      }
    }
    res
  }

  def readTendencyMatrix() {
    var lines = Source.fromFile(matrixFileName).getLines()
    for (i <- 0 until nDis) {
      priorProb(i) = lines.next().toDouble
    }

    for (i <- 0 until nSympt) {
      lines.next() //consume empty line
      for (j <- 0 until nDis) {
        tendencyMatrix(i)(j) = lines.next().toDouble
      }
    }
  }

  def L1(di: Set[Int]): Double = {
    var l1 = 1.0
    for (i <- mPlus) {
      var dProd = 1.0
      for (j <- di) {
        dProd *= (1.0 - tendencyMatrix(i)(j))
      }
      l1 *= (1.0 - dProd)
    }
    l1
  }

  def L2(di: Set[Int]): Double = {
    var absentSymptoms = (0 until nSympt).toSet &~ mPlus //correct

    var l2 = 1.0
    for (j <- di) {
      var mProd = 1.0
      for (l <- absentSymptoms) {
        if (tendencyMatrix(l)(j) > 1.0e-7) {
          mProd *= (1.0 - tendencyMatrix(l)(j))
        }
      }
      l2 *= mProd
    }
    l2
  }

  def L3(di: Set[Int]): Double = {
    var l3 = 1.0
    for (j <- di) {
      l3 *= (priorProb(j) / (1.0 - priorProb(j)))
    }
    l3
  }
}

File: pso_java/MFDFitnessFunction.java
import java.nio.file.*;
import java.nio.charset.Charset;
import java.io.BufferedReader;
import java.io.IOException;


public class MFDFitnessFunction implements FitnessFunction {
 
    byte[] mPlus;
    double[] priorProb;
    double[][] tendencyMatrix;
    final double threshold = 1.0e-7;
    int nSympt;

    public MFDFitnessFunction (String matFileName, int nSympt, int nDis) {
        this.nSympt = nSympt;
        this.priorProb = new double[nDis];
        this.tendencyMatrix = readTendencyMatrix(matFileName, nDis, nSympt);
    }

    private double[][] readTendencyMatrix(String name, int nDis, int nSympt) {
        Path p = FileSystems.getDefault().getPath(name);
        double[][] tendencyMatrix = new double[nSympt][nDis];
        Charset charset = Charset.forName("US-ASCII");
        try (BufferedReader in = Files.newBufferedReader(p, charset)) {
            for (int i = 0; i < nDis; i++) 
                this.priorProb[i] = Double.parseDouble(in.readLine());
            for (int i = 0; i < nSympt; i++) { 
                in.readLine(); //consume empty line
                for (int j = 0; j < nDis; j++) {
                    tendencyMatrix[i][j] = Double.parseDouble(in.readLine());
                }
            }
        } catch(IOException e) {
            e.printStackTrace();
            System.exit(-1);
       }
        return tendencyMatrix;
    }

    private byte[] intToByteArray(int n, int nSympt) {
        String s = Integer.toBinaryString(n);
       
        while(s.length() < nSympt) s = "0" + s;
        byte[] res = new byte[nSympt];
        for (int i = 0; i < nSympt; i++) {
            if (s.charAt(i) == '1') res[i] = 1;
            else res[i] = 0;
        }
        return res;
    }

    public void setMPlus(int mPlus) { this.mPlus = intToByteArray(mPlus, nSympt); }

    public double calculateFitness (byte[] g) {
        return l1(g) * l2(g) * l3(g);
    }

    public double l1 (byte[] g) {
        double l1 = 1.0;
        for (int i = 0; i < mPlus.length; i++) {
            double dProd = 1.0;
            if (mPlus[i] == 1) {
                for (int j = 0; j < g.length; j++){ 
                    if (g[j] == 1) {
                        dProd *= (1.0 - tendencyMatrix[i][j]);
                    }
                }
                l1 *= (1 - dProd);
            }
        }
        return l1;
    }

    public double l2 (byte[] g) {
        double l2 = 1.0;
        for (int j = 0; j < g.length; j++) {
            double mProd = 1.0;
            if (g[j] == 1) { 
                for (int l = 0; l < mPlus.length; l++) {  
                    if ((mPlus[l] == 0) && (tendencyMatrix[l][j] > threshold)) {
                        mProd *= (1.0 - tendencyMatrix[l][j]);
                    }
                }
                l2 *= mProd;
            }            
        }
        return l2;
    }
    
    public double l3 (byte[] g){
        double l3 = 1.0;
        for (int j = 0; j < g.length; j++) {
            if (g[j] == 1){
                l3 *= (priorProb[j] / (1.0 - priorProb[j]));
            }
        }
        return l3;
    }
}
